snapshot/v3_snapshot_test.go

```diff
@@ -0,0 +1,276 @@
+
+	cfg := embed.NewConfig()
+	cfg.Name = "s1"
+	cfg.InitialClusterToken = testClusterTkn
+	cfg.ClusterState = "existing"
+	cfg.LCUrls, cfg.ACUrls = cURLs, cURLs
+	cfg.LPUrls, cfg.APUrls = pURLs, pURLs
+	cfg.InitialCluster = fmt.Sprintf("%s=%s", cfg.Name, pURLs[0].String())
+	cfg.Dir = filepath.Join(os.TempDir(), fmt.Sprint(time.Now().Nanosecond()))
+
+	sp := NewV3(nil, logger.NewPackageLogger(capnslog.NewPackageLogger("github.com/coreos/etcd", "snapshot")))
+
+	err := sp.Restore(dbPath, RestoreConfig{})
+	if err.Error() != `couldn't find local name "" in the initial cluster configuration` {
+		t.Fatalf("expected restore error, got %v", err)
+	}
+	var iURLs types.URLsMap
+	iURLs, err = types.NewURLsMap(cfg.InitialCluster)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = sp.Restore(dbPath, RestoreConfig{
+		Name:                cfg.Name,
+		OutputDataDir:       cfg.Dir,
+		InitialCluster:      iURLs,
+		InitialClusterToken: cfg.InitialClusterToken,
+		PeerURLs:            pURLs,
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	var srv *embed.Etcd
+	srv, err = embed.StartEtcd(cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		os.RemoveAll(cfg.Dir)
+		srv.Close()
+	}()
+	select {
+	case <-srv.Server.ReadyNotify():
+	case <-time.After(3 * time.Second):
+		t.Fatalf("failed to start embed.Etcd")
+	}
+
+	var cli *clientv3.Client
+	cli, err = clientv3.New(clientv3.Config{Endpoints: []string{cfg.ACUrls[0].String()}})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cli.Close()
+	for i := range kvs {
+		var gresp *clientv3.GetResponse
+		gresp, err = cli.Get(context.Background(), kvs[i].k)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if string(gresp.Kvs[0].Value) != kvs[i].v {
+			t.Fatalf("#%d: value expected %s, got %s", i, kvs[i].v, string(gresp.Kvs[0].Value))
+		}
+	}
+}
+
+// TestSnapshotV3RestoreMulti ensures that multiple members
+// can boot into the same cluster after being restored from a same
+// snapshot file.
+func TestSnapshotV3RestoreMulti(t *testing.T) {
+	kvs := []kv{{"foo1", "bar1"}, {"foo2", "bar2"}, {"foo3", "bar3"}}
+	dbPath := createSnapshot(t, kvs)
+	defer os.RemoveAll(dbPath)
+
+	clusterN := 3
+	cURLs, _, srvs := restoreCluster(t, clusterN, dbPath)
+	defer func() {
+		for i := 0; i < clusterN; i++ {
+			os.RemoveAll(srvs[i].Config().Dir)
+			srvs[i].Close()
+		}
+	}()
+
+	// wait for leader election
+	time.Sleep(time.Second)
+
+	for i := 0; i < clusterN; i++ {
+		cli, err := clientv3.New(clientv3.Config{Endpoints: []string{cURLs[i].String()}})
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer cli.Close()
+		for i := range kvs {
+			var gresp *clientv3.GetResponse
+			gresp, err = cli.Get(context.Background(), kvs[i].k)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if string(gresp.Kvs[0].Value) != kvs[i].v {
+				t.Fatalf("#%d: value expected %s, got %s", i, kvs[i].v, string(gresp.Kvs[0].Value))
+			}
+		}
+	}
+}
+
+type kv struct {
+	k, v string
+}
+
+func createSnapshot(t *testing.T, kvs []kv) string {
```

Comment:
createSnapshotFile

probably also comment that the returned string value is the file name